pkg/registry/discovery/endpointslice/strategy.go

```diff
@@ -139,3 +145,37 @@ func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
 		}
 	}
 }
+
+// dropTopologyOnV1 on V1 request wipes the DeprecatedTopology field  and copies
+// the NodeName value into DeprecatedTopology
+func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSlice) {
+	if info, ok := genericapirequest.RequestInfoFrom(ctx); ok {
+
+		// endpointListUpdated should be true for creates
+		endpointListUpdated := true
+		if oldEPS != nil {
+			endpointListUpdated = !apiequality.Semantic.DeepEqual(oldEPS.Endpoints, newEPS.Endpoints)
+		}
+
+		requestGV := schema.GroupVersion{Group: info.APIGroup, Version: info.APIVersion}
+		if requestGV == discoveryv1.SchemeGroupVersion {
+			for i := range newEPS.Endpoints {
+				ep := &newEPS.Endpoints[i]
+
+				// If EndpointList has not been updated, do not clear DeprecatedTopology
+				if endpointListUpdated {
+					//Silently clear out DeprecatedTopology
+					ep.DeprecatedTopology = nil
+				}
+
+				// Set topology[hostname] to match NodeName for compatibility with old clients
+				if ep.NodeName != nil {
+					if ep.DeprecatedTopology == nil {
+						ep.DeprecatedTopology = make(map[string]string, 1)
+					}
+					ep.DeprecatedTopology[corev1.LabelHostname] = *ep.NodeName
+				}
```

Comment:
why isn't this limited to `endpointListUpdated` as well? this means a v1 writer touching only metadata fields (like the garbage collector or something) would cause nodeName to get propagated into deprecatedTopology. is that what we want?

Reply:
I think we need to have nodeName propagated in deprecatedTopology, otherwise older kube-proxies would not be able to read nodeName information since we don't do anything special in the conversion in regards to the hostname topology label. When we remove v1beta1, we could then remove this logic to stop populating topology and eventually deprecatedTopology will be an unused field.

Reply:
> I think we need to have nodeName propagated in deprecatedTopology, otherwise older kube-proxies would not be able to read nodeName information since we don't do anything special in the conversion in regards to the hostname topology label.

When endpoints are being modified via v1, I agree. If the endpoint was written via v1beta1, doing this unconditionally (not just when endpoints were modified via v1) means we'll copy nodeName into topology

Reply:
We still have the version check, but I see your point that it makes sense only to do this when the v1 client is writing to Endpoints